File: 2020/01/Main.kt
import java.io.File
import java.io.InputStream

import tcollier.AocExecutor

fun part1Answer(input: List<Int>): String {
    for (i in 0..input.size-2) {
        for (j in i+1..input.size-1) {
            if (input[i] + input[j] == 2020) {
                return (input[i] * input[j]).toString()
            }
        }
    }
    return "Not Found"
}

fun part2Answer(numbers: List<Int>): String {
    val input = numbers.toMutableList()
    input.sort()
    for (i in 0..input.size-3) {
        var j = i + 1
        var k = input.size - 1
        while (j < k) {
            val sum = input[i] + input[j] + input[k]
            if (sum == 2020) {
                return (input[i] * input[j] * input[k]).toString()
            } else if (sum < 2020) {
                j++
            } else {
                k--
            }
        }
    }
    return "Not Found"
}

fun main(args: Array<String>) {
    val input = mutableListOf<Int>()
    val inputStream: InputStream = File("2020/01/input.txt").inputStream()
    inputStream.bufferedReader().forEachLine { input.add(it.toInt()) }
    val executor = AocExecutor(input, ::part1Answer, ::part2Answer)
    executor.run(args)
}

File: 2020/01/main.rb
require 'set'

require 'aoc_executor'

def pair_with_sum(numbers, sum)
  others = Set.new(numbers[1..-1])
  numbers.each do |number|
    if others.include?(sum - number)
      return [number, sum - number]
    end
  end
  raise "No pair with sum #{sum} found"
end

def triplet_with_sum(numbers, sum)
  numbers.sort!
  (numbers.length - 2).times do |i|
    j = i + 1
    k = numbers.length - 1
    while j < k
      total = numbers[i] + numbers[j] + numbers[k]
      if total == sum
        return [numbers[i], numbers[j], numbers[k]]
      elsif total < sum
        j += 1
      else
        k -= 1
      end
    end
  end
  raise "No triplet with sum #{sum} found"
end

def a(m, n)
  q=p;m.sample(n).tap{|p|p.reduce(:+)==?<.ord.to_s(3).to_i&&q=p} until q;eval(q.join(?*))
end

def part1_ans(input)
  a(input, 2)
end

def part2_ans(input)
  a(input, 3)
end

part1_proc = Proc.new do |input|
  a(input, 2)
end

part2_proc = Proc.new do |input|
  a(input, 3)
end

executor = AocExecutor.new(
  File.readlines(File.join(File.dirname(__FILE__), 'input.txt')).map(&:to_i),
  part1_proc,
  part2_proc
)
executor.run(ARGV)
